### DISCUSSION TITLE
[Request] Example of a "functional"-ish WebAPI leveraging language-ext

### DISCUSSION THREAD

**Discussion by natalie-o-perret** [louthy/language-ext#918] [General]
That would be great if the wiki provides a series of article about to build a project like WebAPI using C# and functional approach with language-ext (eg. about handling errors).

I am sure this could help a lot of folks to get their hands on this library.

**Comment by ErikFeeley** [comment]
I'm really interested in this as well. I'm hoping to work on a sample asp net core API that uses ef core and language-ext this weekend. Worst case scenario at least it might serve as a starting point.

**Comment by natalie-o-perret** [comment]
@ErikFeeley glad to know that I am not all alone :p

**Comment by ErikFeeley** [comment]
Have not had the time to get on this yet : / High level though I want to build a simple Todo api using pretty standard asp net core stuff. Except I want to remove the model binder validation and use language-ext for that. Then also use language-ext for the actual meat of the api and working with EF Core.

**Comment by natalie-o-perret** [comment]
Actually, I have some plans to achieve something similar with a relatively simple kata project following DDD.

I drafted the requirements for a full-stack application below (bigger than a todo api tho):

> Write (little) program that shows how you handle the full stack from front-end to back-end.
> 
> - A user should be able to enter a new flight which has a departure airport and a destination airport.
> - This new flight should be persisted into the system so that it can be retrieved later on.
> - Business logic should be able to calculate the distance between the two aiports their GPS positions and to calculate the fuel needed for this flight, given the aircraft fuel consumption per distance / flight time + takeoff effort.
> - The program should be able to list the different flights which are loaded from a persistence medium, technology for this can be chosen freely.
> - It should also be possible to edit a flight and persist these changes.
> - A report page should be included that shows a summary of all entered data, including calculated data.
>
> Please note that you are free to use the front-end framework (vanilla js, angular, vue, react or even Blazor!) that suits you the best as well as the persistence and infrastructure mediums, be it in memory DB, docker, vault for secrets and... whatnot :-)
> Also a CQRS ES approach is preferred but not mandatory.
> **Please try to follow the holy TDD / BDD / DDD trinity when implementing.**
>

**Comment by sdedalus** [comment]
I built the start of the example you're describing for a talk I gave a few
months ago.  I didn't use language-ext but it shouldn't be a problem to
port the code to use those primitives.
https://github.com/sdedalus/FPTalk1

**Comment by Lonli-Lokli** [comment]
This application has been created by @blakeSaucier https://github.com/louthy/language-ext/pull/637

I believe it has been inspired by Clean Architecture (see this https://paulovich.net/clean-architecture-essentials/ or this https://youtu.be/5OtUm1BLmG0 
 for details)

**Comment by KiritchoukC** [comment]
If anyone is interested, I started a Todo Web Api with LanguageExt.
In .Net 5 and C# 9.
[https://github.com/KiritchoukC/FunctionalArchitectureSample](https://github.com/KiritchoukC/FunctionalArchitectureSample)

I'd like to make it as good as possible so feel free to come and contribute 🙂

**Comment by Lonli-Lokli** [comment]
@KiritchoukC Why everyone is so in love with MediatR?

**Comment by KiritchoukC** [comment]
@Lonli-Lokli I'm not, It's the first time I use it.
What's wrong with MediatR?

**Comment by Lonli-Lokli** [comment]
@KiritchoukC First of all it's not too fast in prod as it used Activator. Second, it leads to a lot of boilerplate (while of course gives clear vision on what dependencies are required for action). 
Also as you see in your example you will need to explicitly use await as it's the requirement for handler.
Another option is https://github.com/ardalis/ApiEndpoints

**Comment by KiritchoukC** [comment]
1. I didn't know that. I don't have a lot of knowledge about Activator, I'll make some research.
2.
- I find the boilerplate ok. Tho I agree, it's a bit overkill for a ToDo Api.
- Oh I can't return the Task without awaiting it?
3. thanks for your recommendation, I'll check it out.
